### PR TITLE
Adding port file for v8

### LINKED ISSUE #372: Add Chrome V8 port
It would be nice if someone add port to Google Chrome V8 JavaScript engine: https://developers.google.com/v8/

### PR DESCRIPTION
**Pull request by tkerwin**
v8 has a quite convoluted build system. 
This is the general sequence of events:
1. Fetch and unzip Google's "depot_tools"
2. Update depot_tools
3. Pull v8 code via depot_tools
4. Generate build framework via depot_tools
5. Append commands to build config file
6. Call ninja to perform the build
7. Install header and library files

Make sure to initialize the Visual Studio command line environment before running `vcpkg install v8:x64-windows` i.e. run `"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Auxiliary\Build\vcvars64.bat"` or the equivalent on your system

Fixes #372

### CODE REVIEW

**Comment by windoze** [comment]
Hi, I've tried your port, but there're some issues.
1. It complained about the depot_tools.zip hash mismatch, I guess Chromium team just made some updates.
2. It complained that it cannot find git
CMake Error at C:/tools/vcpkg/scripts/cmake/vcpkg_execute_required_process.cmake:44 (message):
    Command failed: GIT_EXECUTABLE-NOTFOUND;checkout;branch-heads/6.3
    Working Directory: C:/tools/vcpkg/buildtrees/v8/src/v8/v8
    See logs for more information:
      C:\tools\vcpkg\buildtrees\v8\x64-windows-out.log
      C:\tools\vcpkg\buildtrees\v8\x64-windows-err.log

Call Stack (most recent call first):
  C:/tools/vcpkg/ports/v8/portfile.cmake:69 (vcpkg_execute_required_process)
  C:/tools/vcpkg/scripts/ports.cmake:72 (include)

2 logs files are empty, and I'm pretty sure Git.exe is installed and on the path.

**Comment by windoze** [comment]
Here is the full output:
```
vcpkg install v8:x64-windows

The following packages will be built and installed:
    v8:x64-windows
Starting package 1/1: v8:x64-windows
Building package v8:x64-windows...
-- CURRENT_INSTALLED_DIR=C:/tools/vcpkg/installed/x64-windows
-- DOWNLOADS=C:/tools/vcpkg/downloads
-- CURRENT_PACKAGES_DIR=C:/tools/vcpkg/packages/v8_x64-windows
-- CURRENT_BUILDTREES_DIR=C:/tools/vcpkg/buildtrees/v8
-- CURRENT_PORT_DIR=C:/tools/vcpkg/ports/v8/.
-- Using cached C:/tools/vcpkg/downloads/depot_tools.zip
-- Testing integrity of cached file...
-- Testing integrity of cached file... OK
-- Extracting done
-- Updating depot_tools...
-- Switching to checkout of v8 version 6.3
-- Could NOT find Git (missing: GIT_EXECUTABLE)
CMake Error at C:/tools/vcpkg/scripts/cmake/vcpkg_execute_required_process.cmake:44 (message):
    Command failed: GIT_EXECUTABLE-NOTFOUND;checkout;branch-heads/6.3
    Working Directory: C:/tools/vcpkg/buildtrees/v8/src/v8/v8
    See logs for more information:
      C:\tools\vcpkg\buildtrees\v8\x64-windows-out.log
      C:\tools\vcpkg\buildtrees\v8\x64-windows-err.log

Call Stack (most recent call first):
  C:/tools/vcpkg/ports/v8/portfile.cmake:69 (vcpkg_execute_required_process)
  C:/tools/vcpkg/scripts/ports.cmake:72 (include)


Error: Building package v8:x64-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: v8:x64-windows
  Vcpkg version: 0.0.100-2017-12-09-9f0d33b6ad3af842b133aafa8be1bbf63e08201d

Additionally, attach any relevant sections from the log files above.
```


```
where git
C:\Program Files\Git\cmd\git.exe
```
Seems it failed to find git.

**Comment by xoviat** [comment]
Can you edit your initial post to contain "Fixes #372"?

**Comment by windoze** [comment]
It's weird that suddenly FindGit worked, but there was a new issue:
```
vcpkg install v8:x64-windows
The following packages will be built and installed:
    v8:x64-windows
Starting package 1/1: v8:x64-windows
Building package v8:x64-windows...
-- CURRENT_INSTALLED_DIR=C:/tools/vcpkg/installed/x64-windows
-- DOWNLOADS=C:/tools/vcpkg/downloads
-- CURRENT_PACKAGES_DIR=C:/tools/vcpkg/packages/v8_x64-windows
-- CURRENT_BUILDTREES_DIR=C:/tools/vcpkg/buildtrees/v8
-- CURRENT_PORT_DIR=C:/tools/vcpkg/ports/v8/.
-- Using cached C:/tools/vcpkg/downloads/depot_tools.zip
-- Testing integrity of cached file...
-- Testing integrity of cached file... OK
-- Extracting source C:/tools/vcpkg/downloads/depot_tools.zip
-- Extracting done
-- Updating depot_tools...
-- Initial fetch of v8 git repository
-- Generating build directory x64.optdebug
-- Setting configuration for shared library build
-- Building x64.optdebug through ninja
-- Generating build directory x64.release
-- Setting configuration for shared library build
-- Building x64.release through ninja
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/share/v8/copyright
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include/libplatform
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include/libplatform/libplatform-export.h
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include/libplatform/libplatform.h
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include/libplatform/v8-tracing.h
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include/v8-debug.h
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include/v8-inspector-protocol.h
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include/v8-inspector.h
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include/v8-platform.h
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include/v8-profiler.h
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include/v8-testing.h
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include/v8-util.h
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include/v8-value-serializer-version.h
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include/v8-version-string.h
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include/v8-version.h
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include/v8.h
-- Installing: C:/tools/vcpkg/packages/v8_x64-windows/include/v8config.h
CMake Error at C:/tools/vcpkg/ports/v8/portfile.cmake:147 (file):
  file INSTALL cannot find
  "C:/tools/vcpkg/buildtrees/v8/src/v8/v8/out.gn/x64.release/v8.dll".
Call Stack (most recent call first):
  C:/tools/vcpkg/scripts/ports.cmake:72 (include)


Error: Building package v8:x64-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: v8:x64-windows
  Vcpkg version: 0.0.100-2017-12-09-9f0d33b6ad3af842b133aafa8be1bbf63e08201d

Additionally, attach any relevant sections from the log files above.```

x64-windows-err.log:
```
c:\tools\vcpkg\buildtrees\v8\src\v8\v8\third_party\_gclient_gittmp_instrumented_librariesy9byp3\shallow - The process cannot access the file because it is being used by another process.
```

**Comment by windoze** [comment]
ok, I'v found the problem.
For some reason, ninja didn't build "v8.dll" when it's building "v8-shell.exe", I added "v8.dll" target in portfile.cmake:103 and portfile.cmake:133 and the issue's gone.

**Comment by tkerwin** [comment]
I haven't tried static build, but I'll give it a shot.  
The requirement on `native_blob.bin` and `snapshot.bin` should be removed by the line:
`file(APPEND ${SOURCE_PATH}/v8/v8/out.gn/${BUILDTYPE}/args.gn "v8_use_external_startup_data = false" )`

At least, that worked for me.

**Comment by tkerwin** [comment]
I think maybe the depot tools update affected the v8.dll build issue.  I wasn't seeing it before, and now I am.

**Comment by tkerwin** [comment]
I think the new depot_tools might have changed something with the `build.ninja` generation.  For the `x64.optdebug` build, it has `build v8: phony ./v8.dll.lib`, but in `x64.release`, it has `build v8: phony obj/v8.stamp`.  The dlls are built correctly in the first case but not for release.

**Comment by windoze** [comment]
I see the problem, in portfile.cmake:95, you forgot to add a "\n" at the line end, so the generated args.gn was like:
```
...
visual_studio_version = "2017"v8_use_external_startup_data = false
```

**Comment by tkerwin** [comment]
The latest fixes work better for me.  the v8.dll problem seemed to be some strange side effect of them updating the depot_tools, so I generated the build files using a different method.

**Review by icedream2linxi — COMMENTED** [review]
Do you consider Linux and MacOS?

>是否考虑下 Linux 和 MacOS?

File: ports/v8/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+	vcpkg_execute_required_process(
+	    COMMAND fetch.bat --no-history v8
+	    WORKING_DIRECTORY ${SOURCE_PATH}/v8
+	    LOGNAME ${TARGET_TRIPLET}
+	)
+endif()
+
+message(STATUS "Switching to checkout of v8 version 6.3")
+
+find_package(Git)
```
Comment (icedream2linxi) [inline ports/v8/portfile.cmake]: Use `find_program(GIT NAMES git git.cmd)`, for example `vcpkg_apply_patches.cmake`.

> 使用 `find_program(GIT NAMES git git.cmd)`，例如`vcpkg_apply_patches.cmake`。

**Comment by ivanmoskalev** [comment]
Is there any particular reason why this PR is suspended?
I need v8 in my project, and I'd prefer to add it through v8, so I might try and help the author to merge this.

**Pull request closed without merging**